File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        Set<Integer> set = new HashSet<>();

        for (int i=0; i<topping.length; i++) {
            if (map.containsKey(topping[i])) {
                int temp = map.get(topping[i]);
                map.put(topping[i], temp+1);
            } else {
                map.put(topping[i], 1);
            }
        }

        for (int i=0; i<topping.length-1; i++) {
            set.add(topping[i]);
            int temp = map.get(topping[i]);
            if (temp == 1) {
                map.remove(topping[i]);
            } else {
                map.put(topping[i], temp-1);
            }
            answer += set.size()==map.size() ? 1 : 0;
        }

        return answer;
    }
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    var answer = 0;
    const map = new Map();
    for(const i of topping){
        if(!map.has(i)) map.set(i,1);
        else{
            map.set(i,map.get(i) + 1);
        }
    }
    
    const set = new Set();
    for(const i of topping){
        if(map.get(i) === 1){
            map.delete(i);
        }else map.set(i,map.get(i) -1);
        set.add(i);
        if(set.size === map.size) answer++;
    }
    return answer;
}